### ISSUE TITLE
Add agda-pkg

### ISSUE THREAD

**Issue by turion** [NixOS/nixpkgs#87897]
After #76653, agda is working on nixpkgs, with libraries. It would be great if we can add the package manager `agda-pkg`/`apkg`. Possibly we can make it interoperable with the existing agda infrastructure. The grand vision would be that the agda infrastructure in `nixpkgs` can play the role of a curated, compatible package set, similar to what `stackage` means to the Haskell environment.

* homepage URL: https://github.com/agda/agda-pkg

**Comment by turion** [comment]
I could try and tackle this issue.

@alexarice are you interested as well?

**Comment by alexarice** [comment]
I don't believe that agda-pkg has any sort of version control. In particular most the commits it retrieves won't work on Agda 2.6.1

**Comment by alexarice** [comment]
I could be wrong though

**Comment by turion** [comment]
> I don't believe that agda-pkg has any sort of version control. In particular most the commits it retrieves won't work on Agda 2.6.1

Yes, I think you're right (don't know for sure though). It uses the package index under https://github.com/agda/package-index/, which doesn't have anything like a mutually compatible package set, or dependency resolution.

**Comment by turion** [comment]
I wonder whether I ping relevant Agda ecosystem developers like @jonaprieto here ;)

**Comment by alexarice** [comment]
Ideally we would use a script that updates the agda packages using package-index, though this will generate a load of broken packages at the moment

**Comment by turion** [comment]
> Ideally we would use a script that updates the agda packages using package-index,

Fair enough, and never use `apkg` directly?

> though this will generate a load of broken packages at the moment

True. We can then open upstream issues and hope they'll get resolved eventually

**Comment by alexarice** [comment]
> Fair enough, and never use `apkg` directly?

This makes sense to me

> True. We can then open upstream issues and hope they'll get resolved eventually

I still think for the moment we should pin some commits, but I agree that the lack of versions on libraries is an upstream issue

**Comment by alexarice** [comment]
I am not sure if it is worth doing something like this at the moment when there will likely only be 3-4 working packages

**Comment by turion** [comment]
> I am not sure if it is worth doing something like this at the moment when there will likely only be 3-4 working packages

True, most packages will be broken. Nevertheless I want to give it a shot. It could help to steer the updates of the libraries to 2.6.1. Uploading a PR in a moment.

**Comment by alexarice** [comment]
Could you tag me in the PR?

**Comment by turion** [comment]
https://github.com/NixOS/nixpkgs/pull/87903

**Comment by jonaprieto** [comment]
Hi, I'm not NixOS user, so I might not fully understand this discussion. 

> I don't believe that agda-pkg has any sort of version control. In particular most the commits it retrieves won't work on Agda 2.6.1

@alexarice Do you mean that the libraries in the package-index (their commits) do not compile at the moment with Agda v2.6.1? 

I don't know, that's probably true. Agda changes quite often. In the package-index, we have a few developments in Agda with no specific agda version supported. However that might change in the future, when we have something similar to cabal files. The first attempt could be use
a more detailed library file ".agda-pkg". See https://agda.github.io/agda-pkg/#/?id=agda-pkg-library-file

What version control do you mean? 
Apkg can install the version you need of library. You can use as the source, the package-index, or the url of a git repository. If the package-index gets some update, next time you use apkg, it will let you know you need to update the database. It can also update all the libraries.
So, at least you could say, there is a control over the library versions. 

> which doesn't have anything like a mutually compatible package set, or dependency resolution.

@turion apkg doesn't have "mutually compatible package set" but it has a basic dependency resolution. It will install any library from the field "depend" in the ".agda-lib" file. However, it will only install libraries already indexed in the package-index. However, there is no logic to solve
"version" conflicts between the dependencies. I didn't solve that, based on my experience with Agda. There is no such a need at the moment. It's not Haskell. For Agda, there are just a few packages.

> I am not sure if it is worth doing something like this at the moment when there will likely only be 3-4 working packages

Yes, as I said, there are just a few of packages, so I may agree on this.

**Comment by alexarice** [comment]
To be honest I'm not sure exactly what is meant by version control, The main problem for me is that most libraries have not released a version since 2.6.1 and so will not work if we pull versions from the package index (the way I understand it is that the last version of every library is on the package index?)

**Comment by jonaprieto** [comment]
> (the way I understand it is that the last version of every library is on the package index?)

Yes, at least the one the people use the most, i.e. the standard-library. For the other, it's a bit complicated. Some packages don't even  have a release or version on their source. So, the sha from the commit counts as the version on the database.

**Comment by alexarice** [comment]
Think this can be closed

**Comment by turion** [comment]
@alexarice Yes. I might return to https://github.com/NixOS/nixpkgs/pull/87903 eventually, or open a new issue about using `agda-pkg` at some point.
